Added Practical Examples & TOC to README.md

thecodeclimber opened the pull request:
This is an attempt to make the README.md file more user friendly for people who are not as familiar with eslint and how it works.  I have also added a Table of Contents for improved usability.  

NOTE:  I am new to eslint and estlint-config-prettier and eslint-plugin-prettier.  As such, there may be mistakes in my examples.  If so, please let me know and I will fix them (and, if need be, offer an explanation in the documentation so that other people won't make those mistakes).

lydell commented:
Hi!

Here’s some feedback. The text you’ve added is very long, so please forgive me if I missed obvious things while skimming it.

- It’s too long. I don’t think people will read it.
- Some of your new text duplicates things that are already in the readme, such as explaining what `plugin:prettier/recommended` does?
- I don’t think it makes sense to explain so much about how ESLint plugins and configs work. It should not be up to every plugin/config out there to explain this every time. I consider it a failure on ESLint’s side to have made a plugin/config system that requires so much understanding. Is there some ESLint documentation we can link to instead?
- I don’t understand why we need React and Airbnb examples. The Prettier setup does not change anything from using those plugins/configs.

I think what you’ve written would serve better as a blog post (yet another one on the topic).

R167 commented:
Quick 2 cents: I was actually just trying to understand how this all works and the writeup here was definitely useful. I agree it feels too long for going into the README directly (esp all the stuff about how exactly ESLint works), but not being too familiar with that, I was still happy for the review.

Would it maybe make sense to put this all in a Wiki page or maybe a `CONFIG_DETAILS.md` file that is linked to from the README as "For more info, read this"?

lydell commented:
@R167 Which part of the added text helped you the most?

R167 commented:
@lydell The particularly helpful bit was explaining how `plugin:prettier/recommend` also automatically includes the `eslint-config-prettier` (if installed). I originally wasn't very sure about how this quite worked and was formatting it with both the `prettier/recommended` and the `prettier` (config) line.

It also seemed to clarify a little bit _why_ the plugin needs to be the last extend statement (I skipped over this a couple times when first going through the README, although a more prominent "**NOTE:** `plugin:prettier/recommended` must be the last extends line" would also be sufficient).

lydell commented:
@R167 Thanks! It sounds like the current documentation already includes the information needed the, but just needs some tweaks to make some details more prominent?

https://github.com/prettier/eslint-plugin-prettier#recommended-configuration

Plus saying _why_ to put it last. Is eslint-config-prettier’s docs good enough on that part so we can copy that?

https://github.com/prettier/eslint-config-prettier#installation

thecodeclimber commented:
Thanks for your feedback.  Please see my responses below:

> * It’s too long. I don’t think people will read it.

I think the suggestion of R167 of creating a wiki is a good solution.  You could simply have a Practical Examples section in the wiki and link to it from the README.md file.  That way, it is available to anyone who needs it, but does not crowd the Readme section.

I should add, though, that I purposely explained things as clearly as I could with step-by-step examples.  I imagine a great deal of people who use this package are not particularly familiar with eslint and just don't know what to do or why they do it.  The goal of my examples is to help those people.  

> * Some of your new text duplicates things that are already in the readme, such as explaining what `plugin:prettier/recommended` does?

If memory serves, I thought it would be better if we moved the explanation to the Practical Examples section and leave the basic recommendation in the Recommended Configuration section.  

For me personally, what I find most helpful is having a Practical Examples section with a Table of Contents where I can quickly go to find out what to do in my particular situation with a basic understanding of why I am doing it.  


> * I don’t think it makes sense to explain so much about how ESLint plugins and configs work. It should not be up to every plugin/config out there to explain this every time. I consider it a failure on ESLint’s side to have made a plugin/config system that requires so much understanding. Is there some ESLint documentation we can link to instead?

I just went through the eslint documentation.  They explain, but I don't think it's the most user-friendly explanation or documentation.  I agree you shouldn't have to explain this.  On the other hand, you may wish to ask how much time you are spending in the issues section and elsewhere dealing with question relating to the fact that Eslint did not explain this well.

Personally speaking, I plan to make a video and upload it to Youtube explaining this for the simple fact that I found none of the other tutorials did that.  If you want, I can inform you when it is up.


> * I don’t understand why we need React and Airbnb examples. The Prettier setup does not change anything from using those plugins/configs.

I added the airbnb example to show how prettier plugin/config works with another library (in this case a rather popular one).  For me, the point is not whether or not it overrides anything from airbnb, but just having a concrete example of how to use it with another library.  For me personally, it was not at all obvious how to combine prettier with airbnb or any other library.  

In terms of react, that is because I believe in the past you separated react out with the esling-config-prettier package.  There are still tutorials online that recommend adding it as follows:  `"extends": [ ..., "prettier/react"]`.  I just want to show that that is no longer the case -- that all you have to do is add `prettier`.

ALL of the examples I gave were from particular pain points that I experienced trying to figure out how to do this.  Indeed, I left out one other example (typescript) because I thought it was not needed - that one could figure it out from the other examples.  

But truth be told, I see no reason why you can't have a wiki page with practical examples for the various popular libraries and plugins (vue, angular, svelte, typescript, etc).  It just makes it easier for people who want to use prettier with eslint if there is a readily available example for the use case that they are considering.  To that extent, I think a Kitch Sink example with many different use cases combined can also be helpful.

lydell commented:
Thanks!

It sounds these actions exist:

1. Move most of this PR to the wiki: https://github.com/prettier/eslint-plugin-prettier/wiki/Practical-Examples
2. Make small tweaks to the readme and link to the wiki page
3. Make a PR to ESLint to improve their docs

lydell commented:
@thecodeclimber Do you still feel like doing this?

thecodeclimber commented:
@lydell 

Yes -- I'm sorry I have not responded.  It just has been a busy time at work.  I can't give a clear timeline, but I do want to take care of it.

hgilpin710 commented on README.md:
```diff
@@ -1,4 +1,22 @@
+- [Contributing](#contributing)
+
+## Overview
 
```
Thank you !!!

thecodeclimber commented:
> @thecodeclimber Do you still feel like doing this?

I have to apologize.  I would like to finish this, but I just don't have the time to do so.  I thought I would be able to get around to it, but I see now that was not realistic and I don't think it will be realistic for the foreseable future.  I'm happy for someone else to use what I have done and repurpose it in whatever way you guys think is best.

Again, sorry I can't do more right now.
